packages/firestore/src/remote/rest_connection.ts

```diff
@@ -0,0 +1,168 @@
+    rpcName: string,
+    req: Req,
+    token: Token | null
+  ): Promise<Resp> {
+    const url = this.makeUrl(rpcName, req);
+
+    // The database and/or parent field is already encoded in URL. Specifying it
+    // again in the body is not necessary in production, and will cause
+    // duplicate field errors in the Firestore Emulator. Let's remove it.
+    const jsonObj = ({ ...req } as unknown) as Indexable;
+    delete jsonObj.parent;
+    delete jsonObj.database;
+
+    logDebug(LOG_TAG, 'Sending: ', url, jsonObj);
+
+    const headers = {};
+    this.modifyHeadersForRequest(headers, token);
+
+    return this.performRPCRequest<Indexable, Resp>(
+      rpcName,
+      url,
+      headers,
+      jsonObj
+    ).then(
+      response => {
+        logDebug(LOG_TAG, 'Received: ', response);
+        return response;
+      },
+      (err: FirestoreError) => {
+        logWarn(
+          LOG_TAG,
+          `${rpcName} failed with error: `,
+          err.message,
+          'url: ',
+          url,
+          'request:',
+          req
+        );
+        throw err;
+      }
+    );
+  }
+
+  invokeStreamingRPC<Req, Resp>(
+    rpcName: string,
+    request: Req,
+    token: Token | null
+  ): Promise<Resp[]> {
+    // The REST API automatically aggregates all of the streamed results, so we
+    // can just use the normal invoke() method.
+    return this.invokeRPC<Req, Resp[]>(rpcName, request, token);
+  }
+
+  abstract openStream<Req, Resp>(
+    rpcName: string,
+    token: Token | null
+  ): Stream<Req, Resp>;
+
+  /**
+   * Modifies the headers for a request, adding any authorization token if
+   * present and any additional headers for the request.
+   */
+  protected modifyHeadersForRequest(
+    headers: StringMap,
+    token: Token | null
+  ): void {
+    headers['X-Goog-Api-Client'] = X_GOOG_API_CLIENT_VALUE;
+
+    // Content-Type: text/plain will avoid preflight requests which might
+    // mess with CORS and redirects by proxies. If we add custom headers
+    // we will need to change this code to potentially use the $httpOverwrite
+    // parameter supported by ESF to avoid	triggering preflight requests.
+    headers['Content-Type'] = 'text/plain';
+
+    if (token) {
+      for (const header in token.authHeaders) {
+        if (token.authHeaders.hasOwnProperty(header)) {
+          headers[header] = token.authHeaders[header];
+        }
+      }
+    }
+  }
+
+  /**
+   * Performs an RPC request using an implementation specific networking layer.
+   */
+  protected abstract performRPCRequest<Req, Resp>(
+    rpcName: string,
+    url: string,
+    headers: StringMap,
+    body: Req
+  ): Promise<Resp>;
+
+  private makeUrl<Req>(rpcName: string, req: Req): string {
+    const urlRpcName = RPC_NAME_URL_MAPPING[rpcName];
+    debugAssert(
+      urlRpcName !== undefined,
+      'Unknown REST mapping for: ' + rpcName
+    );
+    const path = ((req as unknown) as Indexable).parent || this.databaseRoot;
```

Comment:
This seems like it could be more self-contained if the caller passed the path they wanted based on the RPC they were invoking. As it stands this is kind of magical because `Indexable` doesn't say anything about how the type should work.

Reply:
I changed the callsites to pass in the path and also removed the manual deletion of `parent` and `database`.